### DISCUSSION TITLE
Programmatically registering DataLoaders

### DISCUSSION THREAD

**Discussion by ellebrecht** [Netflix/dgs-framework#1730] [General]
I tried registering auto-generated `MappedBatchLoader`s dynamically at runtime according to the instructions in #432, in order to get rid of the static boilerplate code annotated with `@DgsDataLoader`. 
The DataLoaders are sucessfully registered in `com.netflix.graphql.dgs.DgsDataFetchingEnvironment.getDataLoaderRegistry()`. They are never run however, and queries using them always time out. 
I can only suspect it has something to do with scheduling. Anything I'm missing here? 
Also, would you consider offering a way to cleanly register DataLoaders dynamically at runtime through the DGS API?

**Comment by srinivasankavitha** [comment]
What release version of the framework are you using? Could you also try with an older version prior to 8.0.0 release?

We have some basic support for dynamic registration  but don't use this scenario extensively within our organization. We can look into supporting dynamic data loader registration for a future release, likely next quarter.

**Reply by srinivasankavitha** [reply]
Ok, so I investigated a bit and the reason the workaround doesn't work is because we switched to using the ScheduledDataLoaderRegistry under the hood for all, which uses dispatch predicates. The framework handles setting the dispatch predicate to DispatchPredicate.ALWAYS or whatever has been set up by the user. However, since you are bypassing this mechanism, you don't get any additional functionality that the framework provides and therefore need to do that manually like shown below, if dynamically registering it yourself.

```
 ScheduledDataLoaderRegistry scheduledRegistry = (ScheduledDataLoaderRegistry) dfe.getDataLoaderRegistry();
 scheduledRegistry.register("greetings", dataloader, DispatchPredicate.DISPATCH_ALWAYS);
```

This is quite hacky, though since it exposes the internals of how the framework sets up data loaders, but that is unfortunately the only way that you can use this workaround for dynamic data loader registration.

As mentioned, we will look into a cleaner abstraction to improve this, but not something that would be available in the next few releases.

**Reply by ellebrecht** [reply]
Thanks for your answer, this way it works fine now. Although it is somewhat hacky because of the reasons you mentioned, this is okay for us for the time being. I think it would be worthwhile to have an API for this sometime in the future.
We are using the DGS 8.1.1 BTW.
Thanks again for this helpful library and your support!
